src/code/model/changes.ts

```diff
@@ -0,0 +1,609 @@
+import WasabeeLink from "./link";
+import WasabeePortal from "./portal";
+import type WasabeeOp from "./operation";
+import WasabeeMarker from "./marker";
+import WasabeeZone from "./zone";
+
+type Change<T, K extends keyof T> = {
+  id: string | number;
+  props?: Partial<Pick<T, K>>;
+  value?: T;
+  type: "addition" | "edition" | "deletion";
+};
+
+type MarkerChange = Change<
+  WasabeeMarker,
+  | "type"
+  | "zone"
+  | "order"
+  | "assignedTo"
+  | "state"
+  | "comment"
+  | "deltaminutes"
+>;
+type LinkChange = Change<
+  WasabeeLink,
+  | "fromPortalId"
+  | "toPortalId"
+  | "color"
+  | "zone"
+  | "order"
+  | "assignedTo"
+  | "state"
+  | "comment"
+  | "deltaminutes"
+>;
+
+type PortalChange = Change<WasabeePortal, "hardness" | "comment">;
+
+type ZoneChange = Change<WasabeeZone, "name" | "color" | "points">;
+
+type OperationChange = Change<
+  WasabeeOp,
+  "name" | "comment" | "referencetime" | "color"
+> & {
+  type: "edition";
+  portals: PortalChange[];
+  links: LinkChange[];
+  markers: MarkerChange[];
+  zones: ZoneChange[];
+};
+
+/**
+ * @param origin Origin object
+ * @param current Object to compare with `origin`
+ * @param keys Keys of the object to compare the content
+ * @returns subset of `current` that differs from `origin`
+ */
+function computeDiff<T, K extends keyof T>(origin: T, current: T, keys: K[]) {
+  const props: Partial<T> = {};
+  let once = false;
+  for (const key of keys) {
+    if (JSON.stringify(origin[key]) !== JSON.stringify(current[key])) {
+      props[key] = current[key];
+      once = true;
+    }
+  }
+  if (once) return props;
+  return null;
+}
+
+/**
+ * @returns summary of link differences, including the link ID
+ */
+function linkChanges(origin: WasabeeLink, current: WasabeeLink) {
+  const changes: LinkChange = {
+    id: origin.ID,
+    type: "edition",
+    props: computeDiff(origin, current, [
+      "fromPortalId",
+      "toPortalId",
+      "color",
+      "zone",
+      "order",
+      "assignedTo",
+      "comment",
+      "state",
+      "deltaminutes",
+    ]),
+  };
+  return changes;
+}
+
+/**
+ * @returns summary of marker differences, including the marker ID
+ */
+function markerChanges(origin: WasabeeMarker, current: WasabeeMarker) {
+  const changes: MarkerChange = {
+    id: origin.ID,
+    type: "edition",
+    props: computeDiff(origin, current, [
+      "type",
+      /*"portalId",*/ // unlikely because we don't swap marker yet
```

Comment:
I thought we'd added (or were planning on adding) maker swap

Reply:
"yet", also what swaping marker does has to be defined first, 
preserving the link ID on anchor swap allow to track anchor and resolve situation where one moved an anchor while another assigned it
imo, marker don't need this refinement and can change ID on swap